pointcloud/octree/basic.go

```diff
@@ -0,0 +1,171 @@
+
+	switch octree.node.nodeType {
+	case InternalNode:
+		for _, children := range octree.node.tree {
+			totalSize += children.Size()
+		}
+
+	case LeafNodeFilled:
+		totalSize = 1
+
+	case LeafNodeEmpty:
+		totalSize = 0
+	}
+
+	return totalSize
+}
+
+// Set checks if the point to be added is a valid point for the octree to hold based on its center and side length.
+// It then recursively iterates through the tree until it finds the appropriate node to add it too. If the found node
+// contains a point already, it will the node into octants and add both the old and new points to them respectively.
+func (octree *basicOctree) Set(p r3.Vector, d pc.Data) error {
+	if (pc.PointAndData{P: p, D: d} == pc.PointAndData{}) {
+		octree.logger.Debug("no data given, skipping insertion")
+		return nil
+	}
+
+	if !checkPointPlacement(octree.center, octree.side, p) {
+		return errors.New("error point is outside the bounds of this octree")
+	}
+
+	switch octree.node.nodeType {
+	case InternalNode:
+		for _, childNode := range octree.node.tree {
+			if checkPointPlacement(childNode.center, childNode.side, p) {
+				// Iterate through children nodes
+				err := childNode.Set(p, d)
+				if err == nil {
+					// Update metadata
+					octree.meta.Merge(p, d)
+				}
+				return err
+			}
+		}
+		return errors.New("error invalid internal node detected, please check your tree")
+
+	case LeafNodeFilled:
+		_, exists := octree.At(p.X, p.Y, p.Z)
+		if exists {
+			// Update data in point
+			octree.node.point.D = d
+			return nil
+		}
+		err := octree.splitIntoOctants()
+		if err != nil {
+			return errors.Errorf("error in splitting octree into new octants: %v", err)
+		}
+		// No update of metadata as the set call below will lead to the InternalNode case due to the octant split
+		return octree.Set(p, d)
+
+	case LeafNodeEmpty:
+		// Update metadata
+		octree.meta.Merge(p, d)
+		octree.node = newLeafNodeFilled(p, d)
+	}
+
+	return nil
+}
+
+// At traverses the octree to see in a point exists at the specified location. If a point does exist, its data is
+// returned along with true, if one is not then no data is returned and the boolean is returned false.
+func (octree *basicOctree) At(x, y, z float64) (pc.Data, bool) {
+	switch octree.node.nodeType {
+	case InternalNode:
+		for _, child := range octree.node.tree {
+			d, exists := child.At(x, y, z)
+			if exists {
+				return d, true
+			}
+		}
+
+	case LeafNodeFilled:
+		if octree.node.point.P.ApproxEqual(r3.Vector{X: x, Y: y, Z: z}) {
+			return octree.node.point.D, true
+		}
+
+	case LeafNodeEmpty:
+	}
+
+	return nil, false
+}
+
+// Iterate TODO: Implement iterate for octree.
+func (octree *basicOctree) Iterate(numBatches, myBatch int, fn func(p r3.Vector, d pc.Data) bool) {
+}
+
+// MarshalOctree TODO: Implement marshalling for octree.
+func (octree *basicOctree) MarshalOctree() ([]byte, error) {
+	return nil, nil
+}
+
+// NewMetaData returns the metadata for.
+func (octree *basicOctree) NewMetaData() MetaData {
```

Comment:
Does this make sense as a public method on octree? Why would a user want an empty meta-data struct from a potentially filled octree? Seems like this makes more sense as either
1. a private helper method on octrees. 
2. a public function of the package, that is not a method of basicOctrees.

Reply:
I copied the exposure of the [NewMetaData](https://github.com/viamrobotics/rdk/blob/f70bc2997afa9a06846c7e8025d08e4daa51fe9a/pointcloud/pointcloud.go#L56) in `pointcloud.go` as I thought it may be used in certain circumstances, however, I can un-expose it for now and re-expose later if needed

Reply:
NewMetaData is not a member method though in pointcloud.go, it is a function of the package. NewMetaData in octree can a public function, but just don't make it a member method.

Reply:
removed OctreeMetadata. Will reimplement if deemed useful for octree serialization but currently it carries too much duplicate information to exist as is. newMetaData will no long exist as octree.meta is just pc.MetaData